larpandoradlcontent/LArDeepLearning/DeepLearningTrackShowerIdAlgorithm.cc

```diff
@@ -0,0 +1,352 @@
+                {   // Make sure we're looking at a hit in the correct tile
+                    const int pixelX(std::get<2>(pixelMap));
+                    const int pixelZ(std::get<3>(pixelMap));
+
+                    object_creation::CaloHit::Metadata metadata;
+                    // Apply softmax to loss to get actual probability
+                    float probShower = exp(outputAccessor[0][1][pixelX][pixelZ]);
+                    float probTrack = exp(outputAccessor[0][2][pixelX][pixelZ]);
+                    float probNull = exp(outputAccessor[0][0][pixelX][pixelZ]);
+                    if (probShower > probTrack && probShower > probNull)
+                        showerHits.push_back(pCaloHit);
+                    else if (probTrack > probShower && probTrack > probNull)
+                        trackHits.push_back(pCaloHit);
+                    else
+                        otherHits.push_back(pCaloHit);
+                    float recipSum = 1.f / (probShower + probTrack);
+                    // Adjust probabilities to ignore null hits and update calo hit metadata
+                    probShower *= recipSum;
+                    probTrack *= recipSum;
+                    metadata.m_propertiesToAdd["Pshower"] = probShower;
+                    metadata.m_propertiesToAdd["Ptrack"] = probTrack;
+                    const StatusCode &statusCode(PandoraContentApi::CaloHit::AlterMetadata(*this, pCaloHit, metadata));
+                    if (statusCode != STATUS_CODE_SUCCESS)
+                        std::cout << "Cannot set calo hit meta data" << std::endl;
+                }
+            }
+        }
+
+        if (m_visualize)
+        {
+            const std::string trackListName("TrackHits_" + listName);
+            const std::string showerListName("ShowerHits_" + listName);
+            const std::string otherListName("OtherHits_" + listName);
+            PANDORA_MONITORING_API(VisualizeCaloHits(this->GetPandora(), &trackHits, trackListName, BLUE));
+            PANDORA_MONITORING_API(VisualizeCaloHits(this->GetPandora(), &showerHits, showerListName, RED));
+            PANDORA_MONITORING_API(VisualizeCaloHits(this->GetPandora(), &otherHits, otherListName, BLACK));
+        }
+    }
+
+    auto end = std::chrono::steady_clock::now();
+    if (m_profile)
+    {
+        auto diff = end - start;
+        float cpu(std::chrono::duration<double, std::milli>(diff).count());
+        PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), "cpu_tree", "cpu_time", cpu));
+        PANDORA_MONITORING_API(FillTree(this->GetPandora(), "cpu_tree"));
+    }
+
+    if (m_visualize)
+        PANDORA_MONITORING_API(ViewEvent(this->GetPandora()));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode DeepLearningTrackShowerIdAlgorithm::ReadSettings(const TiXmlHandle xmlHandle)
```

Comment:
Typically last member function in class definition and in implementation (just helps to find it quickly).